### ISSUE TITLE
Getting `simplify` and `derivative` ready for production

### ISSUE THREAD

**Issue by josdejong** [josdejong/mathjs#779]
I'm working on getting `simplify` and `derivative` ready for production. They are an awesome addition to math.js and they are mature enough to start using in practice (I suppose there will be refinements in the rules in the future).

Here is what I think needs to be done before we can release this functionality:

- [x] Make the JS syntax consistent (accept `Node` or `string`, always return a `Node`):
  - [x] `math.simplify(string | Node) : Node`
  - [x] `math.derivative(string | Node, string | Node) : Node`
- [x] Make the Expression parser syntax consistent (accept `Node` or `string`, always return a `Node`, no magic special cases):
  - [x] `math.eval('derivative("x^2+x", "x")')` should return a `Node`
  - [x] `math.eval('simplify("2x + 3x")')` should return a `Node`
  - [x] `math.eval('derivative(parse("x^2+x"), parse("x"))')` should return a `Node`
  - [x] `math.eval('simplify(parse("2x + 3x"))')` should return a `Node`
  - [x] drop support for the special case `math.simplify('derivative(x^2, x)')` instead I suggest we *always* simplify the output of `derivative`. 
  - [x] drop support for `math.eval('derivative(x^2+x, x)')` since it conflicts with input like `math.eval('derivative(f, x)')` where `f` was defined like `math.eval('f = parse("x^2+x")')` (see future plans below)
  - [x] drop support for special case like `math.eval('simplify(5+derivative(5/(3x), x))')` (see future plans below).
  - [x] drop support for method `Node.simplify`, favor function `math.simplify` instead (keep parts of the library separated, keep API orthogonal)
- [x] Update unit tests, they output deprecation warnings
- [x] By default, derivative calls simplify at the end. Can be disabled by passing option `{simplify: false}`.
- [x] Expose the default rules of simplify
- [x] Ensure all documentation is up to date
  - [x] Function `derivative`
  - [x] Function `simplify`
  - [x] Document simplify rules syntax
  - [x] Write docs for expression parser
- [x] Create a new example `algebra.js` demonstrating `simplify` and `derivative`

Future plans :

- [ ] This functionality is only complete in combination with a function `integral`. It may not be that much work now since we can use the `derivative` implementation as a basis and basically invert all rules. But let's not await this, I want to get this functionality out :).
- [ ] Add support for symbolic computation in operators/functions in the expression parser:
  ```
  x = parse("x")  // or maybe an alias x = symbol("x")
  f = 2x + x
  dx = derivative(f, x)
  simplify(2x + 3x)
  ```
  This requires all operators and functions to be able to support `Node` as input, like `math.add(2, new SymolNode('x'))` should return an `OperatorNode` containing the tree `2+x`.

I'm halfway now, see the latest changes in the `differentiation_consistent_api` branch. @tetslee I've made some changes and reverted a few changes you did in #734 in order to get a consistent, future proof API. I would love to hear If you think this makes sense or whether I'm overlooking anything here.

**Comment by ericman314** [comment]
Those plans all sound like great ideas to me. 

Unfortunately, I think we will find that implementing `integral` is not quite as simple as just computing the derivative in reverse: https://en.wikipedia.org/wiki/Symbolic_integration

**Comment by josdejong** [comment]
Hm yeah that's true. Well, maybe we can start simple: not an extensive set of rules like `derivative` already has but for example integration of basic polynomials?

**Comment by ericman314** [comment]
OK. We can start simple so we have _something_ to show for `integral`, but a complete implementation will be a long work in progress. `integral` is orders of magnitude more complex than `derivative`. Like reversing a hashing function or something.

**Comment by josdejong** [comment]
Yes I understand. Do you think it would be better to not try to add `integral` at all (not giving users false hope), or start simple and see how far we can go?

Other subject: I've removed the method `Node.simplify()`. It can be convenient but the library so far takes a functional approach (`simplify(parse(..))` instead of `parse(...).simplify()`) so this method is sort of redundant. I would also like to keep the expression parser and CAS separated so you can include/exclude different parts of the library more easily.

**Comment by ericman314** [comment]
Let's try to do `integral`, but keep an issue open about continually improving it.

**Comment by josdejong** [comment]
:+1: my idea too :)

**Comment by josdejong** [comment]
(and let's not wait for it with releasing `derivative` and `simplify`)

**Comment by ericman314** [comment]
OK. I noticed an error in the derivative of `exp(x)`. Should I submit a PR to `differentiation_consistent_api`?

**Comment by josdejong** [comment]
Yes please!

can you wait one second, I'm about to push changes (also with a fix in simplify)

**Comment by josdejong** [comment]
Ok pushed

**Comment by ericman314** [comment]
Also, `simplify(exp(x))` is throwing `Symbol is not a constructor`, do you have any idea what this could be about?

**Comment by josdejong** [comment]
yep, try `simplify('exp(x)')`. It's the limitation that we have in the API right now: only accepts string or Node, doesn't parse "on the fly".

It's a vague error though, lets see whether we can improve that because more people will come across this one I thinkg

**Comment by josdejong** [comment]
I get an error "Error: Undefined symbol x" though, not "Symbol is not a constructor" like you?

**Comment by josdejong** [comment]
It may be an error that I've just fixed in simplify/util.js, do you have the changes I pushed  20 mins ago?

**Comment by ericman314** [comment]
That seems to have been the problem. Still trying to figure out this `git` thing.

**Comment by ericman314** [comment]
> I get an error "Error: Undefined symbol x" though, not "Symbol is not a constructor" like you?

Yes, I get that error now:

```js
math.eval('simplify(x + x)')
  Error: Undefined symbol x
math.eval('simplify("x + x")')
  Node {
    implicit: false,
    op: '*',
    fn: 'multiply',
    args: [ Node { value: '2', valueType: 'number' }, Node { name: 'x' } ] }
```

This probably means that the evaluator tries to find a value for `x` before evaluating `simplify`. Perhaps the evaluator needs to know that it should not try to evaulate the argument to `simplify`?

**Comment by ericman314** [comment]
One way might be to create a `SymbolicFunctionNode` that inherits `FunctionNode`, but doesn't compile and evaluate its arguments. Then `simplify`, `derivative`, and `integral` would be `SymbolicFunctionNodes`.

**Comment by josdejong** [comment]
For now we indeed don't support `math.eval('simplify(x + x)')` where `x` is a symbol. In the end, it will be cool to support this though (second bullet under "Future plans"), so in the expression parser you could enter things like:

```
x = parse("x")  // or maybe an alias x = symbol("x")
f = 2x + x
dx = derivative(f, x)
simplify(2x + 3x)
```
In order to get that working, all operators and functions in math.js must support Nodes (which is not the case right now). To support symbolic computation, a function call like `2*x`:
```
multiply(2, new SymbolNode('x'))
```
should return: 
```
new OperatorNode('*', 'multiply', new ConstantNode(2), new SymbolNode('x'))
```

It may be possible to solve this on an expression parser level by introducing "symbolic nodes". Or there may be a way to automatically extend all operators and functions with support for symbolic computation. I haven't really worked out any solution for this yet.

Some use cases of `rawArgs` are described [in the docs](http://mathjs.org/docs/expressions/customization.html#custom-argument-parsing). If we go solve this on the expression parser level it may come in handy, but I'm not sure - don't yet have the problem/solution clear.

**Comment by josdejong** [comment]
OK then, I think `simplify` and `derivative` are ready for prime time, and all docs/examples are up to date too.

I've merged everything in `develop` now, and plan on doing a release tomorrow night (unless we come across something in the meantime).

**Comment by josdejong** [comment]
I've released v3.9.0 and updated the website, description, docs, examples. Support for symbolic computation is a huge milestone! Congrats, get yourself a beer.
